plugin/src/main/kotlin/org/jlleitschuh/gradle/ktlint/KtlintCheck.kt

```diff
@@ -0,0 +1,128 @@
+import org.gradle.api.tasks.OutputFiles
+import org.gradle.api.tasks.PathSensitive
+import org.gradle.api.tasks.PathSensitivity
+import org.gradle.api.tasks.SkipWhenEmpty
+import org.gradle.api.tasks.TaskAction
+import org.jlleitschuh.gradle.ktlint.reporter.KtlintReport
+import org.jlleitschuh.gradle.ktlint.reporter.ReporterType
+import java.io.File
+import javax.inject.Inject
+
+@CacheableTask
+open class KtlintCheck @Inject constructor(objectFactory: ObjectFactory) : DefaultTask() {
+
+    @get:Classpath
+    val classpath: ConfigurableFileCollection = project.files()
+
+    @get:Internal
+    val sourceDirectories: ConfigurableFileCollection = project.files()
+
+    @get:SkipWhenEmpty
+    @get:PathSensitive(PathSensitivity.RELATIVE)
+    @get:InputFiles
+    val sources: FileTree = sourceDirectories.asFileTree.matching { filterable ->
+        KOTLIN_EXTENSIONS.forEach {
+            filterable.include("**/*.$it")
+        }
+    }
+
+    @get:Input
+    val verbose: Property<Boolean> = objectFactory.booleanProperty()
+    @get:Input
+    val ruleSets: ListProperty<String> = objectFactory.listProperty(String::class.java)
+    @get:Input
+    val debug: Property<Boolean> = objectFactory.booleanProperty()
+    @get:Input
+    val android: Property<Boolean> = objectFactory.booleanProperty()
+    @get:Input
+    val ignoreFailures: Property<Boolean> = objectFactory.booleanProperty()
+    @get:Console
+    val outputToConsole: Property<Boolean> = objectFactory.booleanProperty()
+
+    @get:Internal
+    val reports: Map<ReporterType, KtlintReport> = ReporterType.values().map {
+        it to KtlintReport(objectFactory.booleanProperty().apply { set(false) }, it, newOutputFile())
+    }.toMap()
+
+    @get:Internal
+    val enabledReports
+        get() = reports.filterValues { it.enabled.get() }
+
+    @TaskAction
+    fun lint() {
+        val reportsToProcess = enabledReports.values
+        val ktlintVersion = determineKtlintVersion(classpath)
+        ktlintVersion?.let {
+            if (it < SemVer(0, 10, 0)) {
+                throw GradleException("Ktlint versions less than 0.10.0 are not supported. Detected Ktlint version: $it.")
+            }
+        }
+        ktlintVersion?.let { version ->
+            if (version < SemVer(0, 20, 0)) {
+                reportsToProcess.forEach {
+                    val reportOutputPath = it.outputFile.get().asFile.absolutePath
+                    if (reportOutputPath.contains(" ")) {
+                        throw GradleException(
+                                "The output path passed `$reportOutputPath` contains spaces. Ktlint versions less than 0.20.0 do not support this.")
+                    }
+                }
+            }
+        }
+        project.javaexec {
+            it.classpath = classpath
+            it.main = "com.github.shyiko.ktlint.Main"
+            it.args(sourceDirectories.flatMap { dir -> KOTLIN_EXTENSIONS.map { extension -> "${dir.path}/**/*.$extension" } })
```

Comment:
I'm curious why you don't just use the `sources` property you already have in scope for this?

Reply:
Sources are all the Kotlin files in scope. This is possibly a big argument list. If you think Ktlink can happily cope with having all the to be checked `.kt` files as separate arguments on the commandline than that would be fine. I am somewhat concerned about the length of the command line - especially on Windows.

Reply:
Totally fair.